Add release note for 2.7.4 (#13391)
apache/pulsar: site2/website/release-notes.md

- [C++] Remove namespace check for MultiTopicsConsumerImpl [#9520](https://github.com/apache/pulsar/pull/9520)

### 2.7.4

#### 2021-12-14

#### Broker
- Fix DispatchRateLimiter does not take effect. [#8611](https://github.com/apache/pulsar/pull/8611)
- Fix set-publish-rate when using preciseTopicPublishRateLimiterEnable=true. [#10384](https://github.com/apache/pulsar/pull/10384)

> - fix retention policy in topic policy not work. [#11021](https://github.com/apache/pulsar/pull/11021)

Reviewer: Please keep consistent for the following aspects: 
- verb tense (use **present tense**). reason: https://docs.google.com/document/d/1lc5j4RtuLIzlEYCBo97AC8-U_3Erzs_lxpkDuseU0n4/edit#bookmark=id.e8uqh1awkcnp
- capitalization 
    - use sentence-style capitalization. details: https://docs.google.com/document/d/1lc5j4RtuLIzlEYCBo97AC8-U_3Erzs_lxpkDuseU0n4/edit#bookmark=id.s1g5tosy5w3i
    - capitalize proper nouns correctly. details: https://docs.google.com/document/d/1lc5j4RtuLIzlEYCBo97AC8-U_3Erzs_lxpkDuseU0n4/edit#bookmark=id.je5thtkj42cb

Besides, please pay attention to the following aspects：
- do not use contractions in technical information. reason: https://docs.google.com/document/d/1lc5j4RtuLIzlEYCBo97AC8-U_3Erzs_lxpkDuseU0n4/edit#bookmark=id.qa5rwiehb245

Can you check all occurrences overall and update them? Thanks

Author: All right.

- Allow to configure schema compatibility policy for system topics. [#12598](https://github.com/apache/pulsar/pull/12598)
- Cleanup already deleted namespace topics. [#12583](https://github.com/apache/pulsar/pull/12583)
- Add OpAddEntry to pendingAddEntries after the state check. [#12570](https://github.com/apache/pulsar/pull/12570)
- Cancel scheduled tasks when deleting ManagedLedgerImpl. [#12565](https://github.com/apache/pulsar/pull/12565)
- NPE on OpAddEntry while ManagedLedger is closing. [#12364](https://github.com/apache/pulsar/pull/12364)
- Avoid potentially blocking calls to metadata on critical threads. [#12340](https://github.com/apache/pulsar/pull/12340)
- Fixed used after recycle issue in OpAddEntry. [#12103](https://github.com/apache/pulsar/pull/12103)
- Optimize the memory usage of Cache Eviction. [#12045](https://github.com/apache/pulsar/pull/12045)
- Fix wrong key-hash selector used for new consumers after all the previous consumers disconnected. [#12035](https://github.com/apache/pulsar/pull/12035)
- Only auth_errors should log at error level. [#9325](https://github.com/apache/pulsar/pull/9325)

#### Dependency upgrade

> - Upgrade Log4j2 version to 2.14.0 and replace legacy log4j dependency. [#8880](https://github.com/apache/pulsar/pull/8880)

Reviewer 2: We can remove this one.

Author: done

- Remove Boost::System runtime dependency. [#9498](https://github.com/apache/pulsar/pull/9498)
- Upgrade commons-compress to 1.21. [#11345](https://github.com/apache/pulsar/pull/11345)
- Upgrade Jetty to 9.4.43.v20210629. [#11660](https://github.com/apache/pulsar/pull/11660)
- Add maven.restlet.org repository. [#13248](https://github.com/apache/pulsar/pull/13248)

> - Bump log4j to 2.15.0. [#13226](https://github.com/apache/pulsar/pull/13226)

Reviewer 2: for 2.7.4, it should be 2.16.0?

Reviewer 2: And it's better to create use a new section to highlight the security issue fixes.

Author: Ok.

- [Python] Fix fields that are ignoring the required key argument. [#11508](https://github.com/apache/pulsar/pull/11508)
- [Java] Fix Consumer listener does not respect receiver queue size. [#11455](https://github.com/apache/pulsar/pull/11455)
- [Java] Add a optional params scope for pulsar oauth2 client. [#11931](https://github.com/apache/pulsar/pull/11931)
- [Java] fix seek at batchIndex level receive duplicated messages. [#11826](https://github.com/apache/pulsar/pull/11826)
- [Java] Fixed race condition on multi-topic consumer. [#11764](https://github.com/apache/pulsar/pull/11764)
- [Java] remove consumer reference from PulsarClient on subscription failure. [#11758](https://github.com/apache/pulsar/pull/11758)
- [Java] Fixed block forever bug in Consumer.batchReceive. [#11691](https://github.com/apache/pulsar/pull/11691)
- [Java] Send CloseProducer on timeout. [#13161](https://github.com/apache/pulsar/pull/13161)
- [Pulsar perf] Fix log level config for pulsar-admin, pulsar-client and pulsar-perf. [#12915](https://github.com/apache/pulsar/pull/12915)
- [Java] Remove invalid call to Thread.currentThread().interrupt(). [#12652](https://github.com/apache/pulsar/pull/12652)

> - [C++] Removed usages of boost::regex. [#9533](https://github.com/apache/pulsar/pull/9533)

Reviewer: can you get them organized to improve readability? something like:

```
- [Java] xx
- [Java] xx
...
- [C++] xx
- [C++] xx
...
- [Python] xx
- [Python] xx
...
```

Author: Ok

> #### Test

Reviewer 2: We can remove this section

Author: ok

- [C++] Remove namespace check for MultiTopicsConsumerImpl [#9520](https://github.com/apache/pulsar/pull/9520)

### 2.7.4

#### 2021-12-14

#### Security

> - Upgrade log4j to 2.17.0 - [CVE-2021-45105](https://pulsar.apache.org/blog/2021/12/11/Log4j-CVE/)

Author: done

  . [#13392](https://github.com/apache/pulsar/pull/13392)
- Upgrade commons-compress to 1.21 - CVE-2021-35515, CVE-2021-35516, CVE-2021-35517
  and CVE-2021-36090. [#11345](https://github.com/apache/pulsar/pull/11345)

#### Broker

- Fix DispatchRateLimiter does not take effect. [#8611](https://github.com/apache/pulsar/pull/8611)
- Fix set-publish-rate when using
  preciseTopicPublishRateLimiterEnable=true. [#10384](https://github.com/apache/pulsar/pull/10384)
- Fix retention policy in topic policy not work. [#11021](https://github.com/apache/pulsar/pull/11021)
- Fix GetTopicsOfNamespace with binary lookup service not check
  auth. [#11172](https://github.com/apache/pulsar/pull/11172)
- Fix retention of keys in compaction. [#11287](https://github.com/apache/pulsar/pull/11287)
- Fix using partitioned topic name to get Policy. [#11294](https://github.com/apache/pulsar/pull/11294)

> - Fix npe when unloading persistent partitioned topic. [#11310](https://github.com/apache/pulsar/pull/11310)

Author: done

- Close the replicator and replication client when delete cluster. [#11342](https://github.com/apache/pulsar/pull/11342)
- Remove RateLimiter constructors and replace with builder usage. [#11599](https://github.com/apache/pulsar/pull/11599)
- Add backoff for setting for getting topic policies. [#11574](https://github.com/apache/pulsar/pull/11574)
- Fix decode compression managedLedgerInfo data. [#11569](https://github.com/apache/pulsar/pull/11569)
- Fix some topic policy operation without backoff. [#11560](https://github.com/apache/pulsar/pull/11560)
- Fix data lost when using earliest position to subscribe to a
  topic. [#11547](https://github.com/apache/pulsar/pull/11547)

> - Fix the schema deletion when delete topic with delete schema. [#11501](https://github.com/apache/pulsar/pull/11501)

Author: done

- Do not create system topic for heartbeat namespace. [#11499](https://github.com/apache/pulsar/pull/11499)
- Compress managed ledger info. [#11490](https://github.com/apache/pulsar/pull/11490)
- Add backoff for setting for getting topic policies. [#11487](https://github.com/apache/pulsar/pull/11487)

> - Solving precise rate limiting does not takes effect. [#11446](https://github.com/apache/pulsar/pull/11446)

Author: done

- Fix replicated subscriptions direct memory leak. [#11396](https://github.com/apache/pulsar/pull/11396)

> - Close the replicator and replication client when delete cluster. [#11390](https://github.com/apache/pulsar/pull/11390)

Author: done

> - Invalidate the read handle after all cursors consumed. [#11389](https://github.com/apache/pulsar/pull/11389)

Author: Done

- Producer getting producer busy is removing existing producer from
  list. [#11804](https://github.com/apache/pulsar/pull/11804)
- Handle NPE and memory leak when full key range isn't covered with active
  consumers. [#11749](https://github.com/apache/pulsar/pull/11749)
- Call .release() when discarding entry to prevent direct memory
  leak. [#11748](https://github.com/apache/pulsar/pull/11748)
- Fix the topic in fenced state and can not recover. [#11737](https://github.com/apache/pulsar/pull/11737)
- Support disabling non-TLS service ports. [#11724](https://github.com/apache/pulsar/pull/11724)
- Fix the bug, can not update topic when the update topicName is contained by an existed topic as a
  part. [#11686](https://github.com/apache/pulsar/pull/11686)
- Fix Pulsar didn't respond error messages when throw
  InterceptException. [#11650](https://github.com/apache/pulsar/pull/11650)
- Avoid redundant calls for getting the offload policies from the
  offloader. [#11629](https://github.com/apache/pulsar/pull/11629)
- Optimize ManagedLedger Ledger Ownership Check. [#13222](https://github.com/apache/pulsar/pull/13222)

> - Do not reuse the Failed OpAddEntry object which lead bundle unloading

Reviewer: ```suggestion 
- Do not reuse the Failed OpAddEntry object which leads to bundle unloading
```

- Cancel scheduled tasks when deleting ManagedLedgerImpl. [#12565](https://github.com/apache/pulsar/pull/12565)
- NPE on OpAddEntry while ManagedLedger is closing. [#12364](https://github.com/apache/pulsar/pull/12364)
- Avoid potentially blocking calls to metadata on critical
  threads. [#12340](https://github.com/apache/pulsar/pull/12340)
- Fix used after recycle issue in OpAddEntry. [#12103](https://github.com/apache/pulsar/pull/12103)
- Optimize the memory usage of Cache Eviction. [#12045](https://github.com/apache/pulsar/pull/12045)
- Fix wrong key-hash selector used for new consumers after all the previous consumers
  disconnected. [#12035](https://github.com/apache/pulsar/pull/12035)
- Only auth_errors should log at error level. [#9325](https://github.com/apache/pulsar/pull/9325)

#### Dependency upgrade
- Remove Boost::System runtime dependency. [#9498](https://github.com/apache/pulsar/pull/9498)
- Upgrade Jetty to 9.4.43.v20210629. [#11660](https://github.com/apache/pulsar/pull/11660)
- Add maven.restlet.org repository. [#13248](https://github.com/apache/pulsar/pull/13248)

> - Upgrade Zookeeper version to 3.5.9. [#12981](https://github.com/apache/pulsar/pull/12981)

Reviewer: ```suggestion
- Upgrade ZooKeeper version to 3.5.9. [#12981](https://github.com/apache/pulsar/pull/12981)
```

Author: done

- Remove pulsar-dashboard from the publish process of images. [#12534](https://github.com/apache/pulsar/pull/12534)
- Roll back to using Java 8 for docker images. [#12357](https://github.com/apache/pulsar/pull/12357)
- Upgrade TestContainers to 1.15.1. [#9120](https://github.com/apache/pulsar/pull/9120)

#### Client

- [Java] Refactor seek to reuse common logic. [#9670](https://github.com/apache/pulsar/pull/9120)
- [Java] Process partitioned-topic messages on different
  listener-threads. [#10017](https://github.com/apache/pulsar/pull/10017)
- [Java] Fix Consumer listener does not respect receiver queue
  size. [#11455](https://github.com/apache/pulsar/pull/11455)
- [Java] Add a optional params scope for pulsar oauth2 client. [#11931](https://github.com/apache/pulsar/pull/11931)
- [Java] Fix seek at batchIndex level receive duplicated messages. [#11826](https://github.com/apache/pulsar/pull/11826)

> - [Java] Fixed race condition on multi-topic consumer. [#11764](https://github.com/apache/pulsar/pull/11764)

Reviewer: ```suggestion
- [Java] Fix race condition on multi-topic consumer. [#11764](https://github.com/apache/pulsar/pull/11764)
```

Author: done

- [Java] Remove consumer reference from PulsarClient on subscription
  failure. [#11758](https://github.com/apache/pulsar/pull/11758)

> - [Java] Fixed block forever bug in Consumer.batchReceive. [#11691](https://github.com/apache/pulsar/pull/11691)

Reviewer: ```suggestion
- [Java] Fixeblock forever bug in Consumer.batchReceive. [#11691](https://github.com/apache/pulsar/pull/11691)
```

- [Java] Send CloseProducer on timeout. [#13161](https://github.com/apache/pulsar/pull/13161)
- [Java] Remove invalid call to Thread.currentThread().interrupt()
  . [#12652](https://github.com/apache/pulsar/pull/12652)
- [Python] Fix fields that are ignoring the required key argument. [#11508](https://github.com/apache/pulsar/pull/11508)

> - [C++] Removed usages of boost::regex. [#9533](https://github.com/apache/pulsar/pull/9533)

Reviewer: ```suggestion
- [C++] Remove usages of boost::regex. [#9533](https://github.com/apache/pulsar/pull/9533)
```

Author: done

- [C++] Use sendRequestWithId to add timeout to
  hasMessageAvailable. [#11600](https://github.com/apache/pulsar/pull/11600)
- [Pulsar perf] Fix log level config for pulsar-admin, pulsar-client and
  pulsar-perf. [#12915](https://github.com/apache/pulsar/pull/12915)

#### Functions and Pulsar IO

- Fix source stats exposing empty exceptions list. [#11478](https://github.com/apache/pulsar/pull/11478)

> - Fixes function api can not use authdata to check superuser. [#11418](https://github.com/apache/pulsar/pull/11418)

Reviewer: ```suggestion
- Fix function API can not use authdata to check superuser. [#11418](https://github.com/apache/pulsar/pull/11418)
```

Author: done
